exchanges/kraken/kraken_websocket.go

```diff
@@ -210,280 +196,257 @@ func isAwaitingCancelOrderResponses(requestID int64, success bool) bool {
 			return err
 		}
-		if _, ok := dataResponse[0].(float64); ok {
-			err = k.wsReadDataResponse(dataResponse)
-			if err != nil {
-				return err
-			}
+		if len(dataResponse) < 3 {
+			return fmt.Errorf("websocket data array too short: %s", respRaw)
 		}
-		if _, ok := dataResponse[1].(string); ok {
-			err = k.wsHandleAuthDataResponse(dataResponse)
-			if err != nil {
+
+		// For all types of channel second to last field is the channel Name
+		channelName, ok := dataResponse[len(dataResponse)-2].(string)
+		if !ok {
+			return common.GetTypeAssertError("string", dataResponse[len(dataResponse)-2], "channelName")
+		}
+
+		// wsPair is just used for keying the Subs
+		wsPair := currency.EMPTYPAIR
+		if maybePair, ok2 := dataResponse[len(dataResponse)-1].(string); ok2 {
+			var err error
+			if wsPair, err = currency.NewPairFromString(maybePair); err != nil {
 				return err
 			}
 		}
-	} else {
-		var eventResponse map[string]interface{}
-		err := json.Unmarshal(respRaw, &eventResponse)
+
+		c := k.Websocket.GetSubscription(stream.DefaultChannelKey{Channel: channelName, Currency: wsPair, Asset: asset.Spot})
+		if c == nil {
+			return fmt.Errorf("%w: %s %s %s", stream.ErrSubscriptionNotFound, asset.Spot, channelName, wsPair)
+		}
+
+		return k.wsReadDataResponse(c, dataResponse)
+	}
+
+	var eventResponse map[string]interface{}
+	err := json.Unmarshal(respRaw, &eventResponse)
+	if err != nil {
+		return fmt.Errorf("%s - err %s could not parse websocket data: %s",
+			k.Name,
+			err,
+			respRaw)
+	}
+
+	event, ok := eventResponse["event"]
+	if !ok {
+		return nil
+	}
+
+	switch event {
+	case stream.Pong, krakenWsHeartbeat:
+		return nil
+	case krakenWsCancelOrderStatus:
+		var status WsCancelOrderResponse
+		err := json.Unmarshal(respRaw, &status)
 		if err != nil {
-			return fmt.Errorf("%s - err %s could not parse websocket data: %s",
+			return fmt.Errorf("%s - err %s unable to parse WsCancelOrderResponse: %s",
 				k.Name,
 				err,
 				respRaw)
 		}
-		if event, ok := eventResponse["event"]; ok {
-			switch event {
-			case stream.Pong, krakenWsHeartbeat:
-				return nil
-			case krakenWsCancelOrderStatus:
-				var status WsCancelOrderResponse
-				err := json.Unmarshal(respRaw, &status)
-				if err != nil {
-					return fmt.Errorf("%s - err %s unable to parse WsCancelOrderResponse: %s",
-						k.Name,
-						err,
-						respRaw)
-				}
 
-				success := true
-				if status.Status == "error" {
-					success = false
-					cancelOrdersStatusMutex.Lock()
-					if _, ok := cancelOrdersStatus[status.RequestID]; ok {
-						if cancelOrdersStatus[status.RequestID].Error == "" { // save the first error, if any
-							cancelOrdersStatus[status.RequestID].Error = status.ErrorMessage
-						}
-					}
-					cancelOrdersStatusMutex.Unlock()
+		success := true
+		if status.Status == "error" {
+			success = false
+			cancelOrdersStatusMutex.Lock()
+			if _, ok := cancelOrdersStatus[status.RequestID]; ok {
```

Comment:
Don't need to have 3 consecutive map lookups. You can name the var differently. 
```diff
diff --git a/exchanges/kraken/kraken_websocket.go b/exchanges/kraken/kraken_websocket.go
index d83c84213..5e5f6e52e 100644
--- a/exchanges/kraken/kraken_websocket.go
+++ b/exchanges/kraken/kraken_websocket.go
@@ -257,9 +257,9 @@ func (k *Kraken) wsHandleData(respRaw []byte) error {
                if status.Status == "error" {
                        success = false
                        cancelOrdersStatusMutex.Lock()
-                       if _, ok := cancelOrdersStatus[status.RequestID]; ok {
-                               if cancelOrdersStatus[status.RequestID].Error == "" { // save the first error, if any
-                                       cancelOrdersStatus[status.RequestID].Error = status.ErrorMessage
+                       if ptrToStruct, ok := cancelOrdersStatus[status.RequestID]; ok {
+                               if ptrToStruct.Error == "" { // save the first error, if any
+                                       ptrToStruct.Error = status.ErrorMessage
                                }
                        }
```

Reply:
Sure. I didn't write this code at all, for what it's worth - it was just moved due to indentation fixes.

I'll fix it as a driveby though.

Reply:
Nit - ptrTo is idiosyncratic go naming. 🙂